src/config.cpp

```diff
@@ -0,0 +1,322 @@
+/*************************************************************************
+ * Copyright (C) 2018-2020 Blue Brain Project
+ *                         Jonas Karlsson <jonas.karlsson@epfl.ch>
+ *                         Juan Hernando <juan.hernando@epfl.ch>
+ *
+ * This file is part of 'libsonata', distributed under the terms
+ * of the GNU Lesser General Public License version 3.
+ *
+ * See top-level COPYING.LESSER and COPYING files for details.
+ *************************************************************************/
+
+#include <bbp/sonata/config.h>
+
+#include <algorithm>  // transform
+#include <cassert>
+#include <fstream>
+#include <memory>
+#include <regex>
+#include <set>
+#include <string>
+
+#include <fmt/format.h>
+#include <ghc/filesystem.hpp>
+#include <nlohmann/json.hpp>
+
+#include "population.hpp"
+#include "utils.h"
+
+
+namespace {
+using bbp::sonata::CircuitConfig;
+using bbp::sonata::SonataError;
+
+// to be replaced by std::filesystem once C++17 is used
+namespace fs = ghc::filesystem;
+
+class PathResolver
+{
+  public:
+    PathResolver(const std::string& basePath)
+        : _basePath(fs::absolute(fs::path(basePath))) {}
+
+    std::string toAbsolute(const std::string& pathStr) const {
+        const fs::path path(pathStr);
+        const auto absolute = path.is_absolute() ? path : fs::absolute(_basePath / path);
+        return absolute.lexically_normal().string();
+    }
+
+  private:
+    const fs::path _basePath;
+};
+
+
+std::map<std::string, std::string> replaceVariables(std::map<std::string, std::string> variables) {
+    constexpr size_t maxIterations = 10;
+
+    bool anyChange = true;
+    size_t iteration = 0;
+
+    while (anyChange) {
+        anyChange = false;
+        auto variablesCopy = variables;
+
+        for (const auto& vI : variables) {
+            const auto& vIKey = vI.first;
+            const auto& vIValue = vI.second;
+
+            for (auto& vJ : variablesCopy) {
+                auto& vJValue = vJ.second;
+                const auto startPos = vJValue.find(vIKey);
+
+                if (startPos != std::string::npos) {
```

Comment:
Shouldn't we check that the key is followed by something non-alphanumeric to make sure that the found variable is terminated properly?

Seems that actually `${BASE_DIR}` etc would be better than `$BASE_DIR`.

Reply:
> Shouldn't we check that the key is followed by something non-alphanumeric to make sure that the found variable is terminated properly?

Yeah, good point.  Nice catch.

> Seems that actually ${BASE_DIR} etc would be better than $BASE_DIR.

It's following the spec: https://github.com/AllenInstitute/sonata/blob/master/docs/SONATA_DEVELOPER_GUIDE.md#tying-it-all-together---the-networkcircuit-config-file

Reply:
I see this just below your link:
```
"network": "${BASE_DIR}/circuit_config.json"
```

Seems that the variable replacement should be more than just plain search + replace.

Reply:
Seems fair.